### PR TITLE
Get to the current module's MVID using a mechanism that works for portable as well as desktop

### PR DESCRIPTION
**Pull request by JohnHamby**
@dotnet/roslyn-compiler @dotnet/testimpact @jcouv @AlekseyTs @tannergooding @ManishJayaswal @drognanar

### CODE REVIEW

**Comment by jcouv** [comment]
Btw, John, please don't forget to start a design doc.

**Comment by jcouv** [comment]
Is there a test that would have failed but now passes with this change?

**Comment by jcouv** [comment]
Also, what happens if the types you need are unavailable?
PR #9536 is san example of test that builds a corlib with some types missing.

File: src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs [hunk 1]
```diff
@@ -51,16 +52,23 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
...
                     factory.Property(
                         factory.Property(
-                            factory.TypeOfPrivateImplementationDetails(),
+                            factory.Convert(
+                                factory.WellKnownType(WellKnownType.System_Type),
+                                factory.Call(
+                                    null,
+                                    (MethodSymbol)Binder.GetWellKnownTypeMember(compilationState.Compilation, WellKnownMember.System_Reflection_IntrospectionExtensions__GetTypeInfo, diagnostics, syntax: syntax),
+                                    factory.TypeOfPrivateImplementationDetails()),
+                                ConversionKind.ImplicitReference),
```
Comment (KevinH-MS) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: Do we need to retain a fallback for targeting .NET 2.0?

Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: @KevinH-MS yes. We absolutely must be able to target .NET 2.0 with the compiler.  If there is not a test for that scenario we should add one as a part of making this change to verify thati tstil lworks.

Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: Do we need to be able to target 2.0 when compiling for instrumentation?

Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: Is there a reason to disallow it?  In general the compiler doesn't understand runtimes at alll.  All we understand is type availability.  If the types are available then we copmile.

**Comment by JohnHamby** [comment]
No existing test would fail without the change, because all of our instrumentation tests target desktop. Can someone point me at a test that targets portable, and at one that targets .Net 2.0?

If types/members are not available, the mechanisms for getting to them (WellKnownType and GetWellKnownTypeMember) are supposed to report errors.

File: src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs [hunk 2]
```diff
@@ -51,16 +52,23 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
                 body.Add(payloadInitialization);
             }
 
-            // MVID = TypeOf(PrivateImplementationDetails).Module.ModuleVersionId;
+            // MVID = System.Reflection.IntrospectionExtensions.GetTypeInfo(TypeOf(PrivateImplementationDetails)).Module.ModuleVersionId;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: It would be good to explain what does System.Reflection.IntrospectionExtensions.GetTypeInfo do.

File: src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs [hunk 3]
```diff
@@ -51,16 +52,23 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
...
                 factory.Assignment(
                     factory.ModuleVersionId(),
                     factory.Property(
                         factory.Property(
-                            factory.TypeOfPrivateImplementationDetails(),
+                            factory.Convert(
+                                factory.WellKnownType(WellKnownType.System_Type),
+                                factory.Call(
+                                    null,
+                                    (MethodSymbol)Binder.GetWellKnownTypeMember(compilationState.Compilation, WellKnownMember.System_Reflection_IntrospectionExtensions__GetTypeInfo, diagnostics, syntax: syntax),
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: Unlike GetWellKnownType API, GetWellKnownTypeMember can return null, which, I suspect, won't be handled properly. SyntheticBoundNodeFactory should have a helper that simply takes WellKnownMember value instead of a method symbol and takes care of things.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: We should have a test for this scenario.

File: src/Compilers/Core/Portable/WellKnownMember.cs [hunk 4]
```diff
@@ -67,6 +66,8 @@ internal enum WellKnownMember
 
         System_Reflection_Missing__Value,
 
+        System_Reflection_IntrospectionExtensions__GetTypeInfo,
+        System_Type__Module,
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/WellKnownMember.cs]: I think it would be appropriate to keep System_Type__Module next to other System.Type members.

File: src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs (hunk 1 above)
Comment (tannergooding) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: @JohnHamby, I may be wrong, but I thought we had plans to support instrumentation on legacy applications as well.

Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: I've added logic (and tests for it) to generate a valid initialization of the MVID property when possible on vintage desktop and portable.

**Comment by JohnHamby** [comment]
I've pushed some new commits to address review feedback. By following the example provided by @jcouv , I was able to craft tests to verify expected patterns of IL generation based on what's available in the target framework, as well as tests for proper handling of missing required members.

**Comment by jcouv** [comment]
Reviewing now.

File: src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs [hunk 5]
```diff
@@ -1169,8 +1187,186 @@ .locals init (bool[] V_0)
...
+
+            try
+            {
+                CompilationVerifier verifier = CompileWithNoFramework(source + InstrumentationHelperSource, emitOptions: EmitOptions.Default.WithInstrument("Test.Flag"));
+            }
+            catch (SyntheticBoundNodeFactory.MissingPredefinedMember missing)
+            {
+                Assert.True(missing.Diagnostic.ToString().Contains("IntrospectionExtensions.GetTypeInfo"));
+                return;                
+            }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs]: I'm a little worried about this `catch` because I'm not sure if the exception comes from the test framework or the compiler, and the compiler should not throw in such a case.

Is it possible to use this tests pattern to check the diagnostics instead?

```C#

            var comp = CreateCompilationWithMscorlib(source);
            comp.VerifyDiagnostics(
                // (6,9): error CS8058: Feature 'tuples' is experimental and unsupported; use '/features:tuples' to enable.
                //         (int, int) x = (1, 1);
                Diagnostic(ErrorCode.ERR_FeatureIsExperimental, "(int, int)").WithArguments("tuples", "tuples").WithLocation(6, 9),
                // (6,24): error CS8058: Feature 'tuples' is experimental and unsupported; use '/features:tuples' to enable.
                //         (int, int) x = (1, 1);
                Diagnostic(ErrorCode.ERR_FeatureIsExperimental, "(1, 1)").WithArguments("tuples", "tuples").WithLocation(6, 24)
                );
```

Run it once with `comp.VerifyDiagnostics();` and the output window will give you what you can copy and paste into the parameters.

File: src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs [hunk 6]
```diff
@@ -1169,8 +1187,186 @@ .locals init (bool[] V_0)
...
+
+            try
+            {
+                CompilationVerifier verifier = CompileWithNoFramework(source + InstrumentationHelperSource, emitOptions: EmitOptions.Default.WithInstrument("Test.Flag"));
+            }
+            catch (SyntheticBoundNodeFactory.MissingPredefinedMember missing)
+            {
+                Assert.True(missing.Diagnostic.ToString().Contains("Module.ModuleVersionId"));
+                return;
+            }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs]: Same comment on verifying diagnostics.

**Comment by jcouv** [comment]
LGTM with exception of diagnostic check that should be fixed.

File: src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs [hunk 7]
```diff
@@ -51,22 +52,61 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
...
+            // The Portable mechanism also works on Desktop, but not on versions prior to 4.5, so there is no one single mechanism that work on all platforms.
+
+            BoundExpression moduleReference;
+            Symbol system_Type__Module = compilation.GetWellKnownTypeMember(WellKnownMember.System_Type__Module);
+            if (system_Type__Module != null)
+            {
+                // moduleReference = TypeOf(PrivateImplementationDetails).Module;
+                moduleReference = factory.Property(typeInstance, WellKnownMember.System_Type__Module);
+            }
+
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: Please remove this empty line.

File: src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs [hunk 8]
```diff
@@ -51,22 +52,61 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
...
+            // Getting to the MVID requires getting to a representation of the current module. If the target platform is Desktop, the module is
+            // available as a property of a type instance. If the target platform is Portable, a type instance does not have a module property.
+            // Instead, the module is available as a property of a member info object, which is available as a base type of a type info object,
+            // which is available via a GetTypeInfo extension method.
+            //
+            // The Portable mechanism also works on Desktop, but not on versions prior to 4.5, so there is no one single mechanism that work on all platforms.
+
+            BoundExpression moduleReference;
+            Symbol system_Type__Module = compilation.GetWellKnownTypeMember(WellKnownMember.System_Type__Module);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: It would be good to add a comment that we are relying on SyntheticBoundNodeFactory to report any appropriate use-site diagnostics.

File: src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs [hunk 9]
```diff
@@ -51,22 +52,61 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
...
+
+            else
+            {
+                // moduleReference = ((System.Reflection.MemberInfo)System.Reflection.IntrospectionExtensions.GetTypeInfo(TypeOf(PrivateImplementationDetails))).Module;
+                moduleReference =
+                    factory.Property(
+                        factory.Convert(
+                            factory.WellKnownType(WellKnownType.System_Reflection_MemberInfo),
+                            factory.StaticCall(
+                                factory.WellKnownType(WellKnownType.System_Reflection_IntrospectionExtensions),
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: Why do we need the type for the static call? Wouldn't the member be sufficient?

File: src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs [hunk 10]
```diff
@@ -582,6 +582,15 @@ public BoundExpression StaticCall(TypeSymbol receiver, MethodSymbol method, para
             return Call(null, method, args);
         }
 
+        public BoundExpression StaticCall(TypeSymbol receiver, WellKnownMember method, params BoundExpression[] args)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs]: Please remove the receiver parameter.

File: src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs [hunk 11]
```diff
@@ -582,6 +582,15 @@ public BoundExpression StaticCall(TypeSymbol receiver, MethodSymbol method, para
...
+        {
+            MethodSymbol methodSymbol = WellKnownMethod(method);
+            Debug.Assert(methodSymbol.IsStatic);
+            Debug.Assert(receiver.GetMembers(methodSymbol.Name).OfType<MethodSymbol>().Contains(methodSymbol));
+            Binder.ReportUseSiteDiagnostics(methodSymbol, Diagnostics, Syntax);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs]: Why is this needed? Haven't WellKnownMethod already done this?

File: src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs [hunk 12]
```diff
@@ -1077,11 +1095,290 @@ static void TestMain()
...
+        return 3;
+    }
+}
+";
+
+            try
+            {
+                CompilationVerifier verifier = CompileWithNoFramework(source + InstrumentationHelperSource, emitOptions: EmitOptions.Default.WithInstrument("Test.Flag"));
+            }
+            catch (SyntheticBoundNodeFactory.MissingPredefinedMember missing)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs]: What is throwing this exception? Compiler shouldn't throw in this scenario.

File: src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs [hunk 13]
```diff
@@ -51,22 +52,61 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
...
+            // Initialize the module version ID (MVID) field. Dynamic instrumentation requires the MVID of the executing module, and this field makes that accessible.
+            body.Add(InitializeMVID(factory, diagnostics));
 
             BoundStatement returnStatement = factory.Return();
             body.Add(returnStatement);
 
             factory.CloseMethod(factory.Block(body.ToImmutableAndFree()));
         }
+
+        private BoundStatement InitializeMVID(SyntheticBoundNodeFactory factory, DiagnosticBag diagnostics)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: `static`?

File: src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs (hunk 11 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs]: I constructed the body of this method from a synthesis of the other StaticCall methods and some Property methods. The results looked kind of silly to me, but I wanted to make sure that I wasn't dropping any important checks. I'll streamline this to something rational.

Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs]: The call to ReportUseSiteDiagnostics turns out to be necessary. WellKnownMethod does not report the diagnostics. Maybe it should.

File: src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs (hunk 13 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: Yes, this method can be static. Thanks.

Comment (cston) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: `diagnostics` is not used.

Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs]: Indeed. Apparently the diagnostics parameter became unnecessary after introducing a helper for StaticCall, and I failed to notice this.

**Comment by JohnHamby** [comment]
I have pushed a new commit to address recent feedback. The tests still include exception handlers, but the exceptions are now generated by the testing infrastructure rather than by the compiler.

File: src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs (hunk 11 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs]: Agree, we should clean this up. Separately, I guess.

File: src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs [hunk 14]
```diff
@@ -1281,10 +1282,15 @@ static int TestMain()
...
                 CompilationVerifier verifier = CompileWithNoFramework(source + InstrumentationHelperSource, emitOptions: EmitOptions.Default.WithInstrument("Test.Flag"));
             }
-            catch (SyntheticBoundNodeFactory.MissingPredefinedMember missing)
+            catch (EmitException exception)
             {
-                Assert.True(missing.Diagnostic.ToString().Contains("IntrospectionExtensions.GetTypeInfo"));
-                return;                
+                foreach (Diagnostic diagnostic in exception.Diagnostics)
+                {
+                    if (diagnostic.ToString().Contains("Missing compiler required member 'System.Reflection.IntrospectionExtensions.GetTypeInfo'"))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs]: Formatting diagnostics this way is sensitive to the current UI language. At least consider passing InvariantCulture to the ToString, but using VerifyDagnostics helper is preferred,

**Comment by JohnHamby** [comment]
I have pushed a commit that, thanks to help from @AlekseyTs, removes the fragility of depending on the formatting of diagnostics.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by jcouv** [comment]
LGTM

**Comment by tannergooding** [comment]
Merging at the request of @JohnHamby and @ManishJayaswal to unblock our dogfooding work. It looks like proper sign-off has been attained and all tests are passing.
